Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs

```diff
@@ -114,12 +116,25 @@ cancelRideImpl rideId bookingCReason = do
             && maybe True (\nsBounds -> isJust booking.fareParams.nightShiftCharge == isNightShift nsBounds now) farePolicy.nightShiftBounds
     if isRepeatSearch
       then do
+        blockListedDriverList <- addDriverToSearchCancelledList searchReq.id ride
         driverPoolCfg <- getDriverPoolConfig merchant.id searchReq.estimatedDistance
+        logDebug $ "BlockListed Drivers-" <> show blockListedDriverList
         driverPool <- calculateDriverPool DP.Estimate driverPoolCfg (Just searchTry.vehicleVariant) searchReq.fromLocation merchant.id True Nothing
-        if not (null driverPool)
+        let newDriverPool = filter (\dpr -> cast dpr.driverId `notElem` blockListedDriverList) driverPool
+        logDebug $ "New Driver Pool" <> show newDriverPool
+        if not (null newDriverPool)
           then repeatSearch merchant farePolicy searchReq searchTry booking ride SBCR.ByDriver now driverPoolCfg
           else BP.sendBookingCancelledUpdateToBAP booking merchant bookingCReason.source
       else BP.sendBookingCancelledUpdateToBAP booking merchant bookingCReason.source
+  where
+    addDriverToSearchCancelledList searchReqId ride = do
+      let keyForDriverCancelledList = DP.mkBlockListedDriversKey searchReqId
+      cacheBlockListedDrivers keyForDriverCancelledList ride.driverId
+      Redis.getList keyForDriverCancelledList
+
+    cacheBlockListedDrivers key driverId = do
+      searchRequestExpirationSeconds <- asks (.searchRequestExpirationSeconds)
+      Redis.rPushExp key [driverId] (round searchRequestExpirationSeconds)
```

Comment:
what if ride has been cancelled twice?

Reply:
Two different drivers cancel the ride then both will be blocked. Any driver who cancels will not be considered for the next batch.